sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1059 @@
+          }
+        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        }
+      }
+    }
+
+    /**
+     * Throws exception if you set different InputFormat key or value class than InputFormat's
+     * actual key or value class. If you set incorrect classes then, it may result in an error like
+     * "unexpected extra bytes after decoding" while the decoding process happens. Hence this
+     * validation is required.
+     */
+    private void validateUserInputForKeyAndValue() throws IOException, InterruptedException {
+      ParameterizedType genericClassType = determineGenericType();
+      RecordReader<?, ?> reader = fetchFirstRecordReader();
+      boolean isCorrectKeyClassSet =
+          validateKeyClass(genericClassType.getActualTypeArguments()[0].getTypeName(), keyCoder,
+              reader.getCurrentKey());
+      boolean isCorrectValueClassSet =
+          validateValueClass(genericClassType.getActualTypeArguments()[1].getTypeName(),
+              valueCoder, reader.getCurrentValue());
+      if (!isCorrectKeyClassSet) {
+        Class<?> actualClass =
+            conf.getHadoopConfiguration().getClass(HadoopInputFormatIOConstants.KEY_CLASS,
+                Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_KEY_CLASS_ERROR_MSG, reader
+                .getCurrentKey().getClass().getName(), actualClass.getName()));
+      }
+      if (!isCorrectValueClassSet) {
+        Class<?> actualClass =
+            conf.getHadoopConfiguration().getClass(HadoopInputFormatIOConstants.VALUE_CLASS,
+                Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_VALUE_CLASS_ERROR_MSG, reader
+                .getCurrentValue().getClass().getName(), actualClass.getName()));
+      }
+    }
+
+    /**
+     * Returns true if key class set by the user is compatible with the key class of a pair returned
+     * by RecordReader. User provided key class is validated against the parameterized type's type
+     * arguments of InputFormat. If parameterized type has any type arguments such as T, K, V, etc
+     * then validation is done by encoding and decoding key object of first pair returned by
+     * RecordReader.
+     */
+    private <T> boolean validateKeyClass(String inputFormatGenericClassName, Coder<K> coder,
+        Object keyObject) {
+      try {
+        Class<?> inputClass = Class.forName(inputFormatGenericClassName);
+        /*
+         * Validates key class with InputFormat's parameterized type.
+         */
+        return validateClassesEquality(
+            inputClass,
+            conf.getHadoopConfiguration().getClass(HadoopInputFormatIOConstants.KEY_CLASS,
+                Object.class));
+      } catch (ClassNotFoundException e) {
```

Comment:
instead of catching classNotFoundException here would it be easier if determineGenericType threw and exception at that point and we differentiated between the two scenarios ("extends generic" and "doesn't extend generic") at that point? 

apologies if I don't understand this properly, I'm trying to determine how this works and if we can make it clearer.